viewing_party/party.py

```diff
@@ -1,23 +1,212 @@
+def get_friends_unique_watched(users_data):
+    friends_movies_set = []
+    users_movies_set = []
+    # loop through the friends watched list and add each movie title to the set
+    for friend in users_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies_set.append(movie["title"])
+    # loop through the user's watched list and add each movie title to the set
+    for movie in users_data["watched"]:
+        users_movies_set.append(movie["title"])
+    # turn list to set so we can find the difference between user's and friend's list in O(n) time
+    friends_movies_set = set(friends_movies_set)
+    users_movies_set = set(users_movies_set)
+    unique_movies_titles = friends_movies_set - users_movies_set
+    unique_movies = []
+    # loop through the movies in the friends's watched list and if the movie title is in the set of unique movies but not already in the list, append that movie to a list and return it
+    for friend in users_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_movies_titles and movie not in unique_movies:
+                unique_movies.append(movie)
+    return unique_movies
+        
+
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+# returns a list of recommended movies based on friends watched, available streaming services, and unwatched movies
+# parameters: users_data
+# returns a list of movies objects 
+def get_available_recs(user_data):
+    # use the unique_friends_movies functions to return a list of movies that have been watched by friends but not the user
+    unique_friends_movies = get_friends_unique_watched(user_data)
+    recommendations = []
+    # loop through each movie that friedns have watched but not user and check if the user has access to the host of each movie 
+    for movie in unique_friends_movies:
+        # append to list if movie can be streamed
+        if movie["host"] in user_data["subscriptions"]:
+            recommendations.append(movie)
+    return recommendations
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+# returns a list of recommended movies based on friends watched, most watched genre, and unwatched movies
+# parameters: users_data
+# returns a list of movies objects 
+def get_new_rec_by_genre(user_data):
+    # get user's most watched genre by looping through movies and storing their frequencies in a dictionary
+    genre_freq = {}
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_freq:
+            genre_freq[movie["genre"]] += 1
+        else:
+            genre_freq[movie["genre"]] = 1
+    frequency = 0
+    genre_saved = ""
+    for genre, freq in genre_freq.items():
+        if frequency < freq:
+            frequency = freq
+            genre_saved = genre

```

Comment:
This was a good opportunity to avoid repeating yourself by using the `get_most_watched_genre` function that you defined in wave 2 which does this same thing!

Don't be afraid to use functions from earlier waves as helper functions, that's part of the design of the waves!